tidb_test.go

```diff
@@ -332,18 +333,39 @@ func (s *testMainSuite) TestSchemaValidity(c *C) {
 	err = <-endCh2
 	c.Assert(err, IsNil, Commentf("err:%v", err))
 
-	err = se.Close()
-	c.Assert(err, IsNil)
-	err = se1.Close()
-	c.Assert(err, IsNil)
-	err = se2.Close()
-	c.Assert(err, IsNil)
+	se.Close()
+	se1.Close()
+	se2.Close()
 	sessionctx.GetDomain(ctx).Close()
 	err = store.Close()
 	c.Assert(err, IsNil)
 	localstore.MockRemoteStore = false
 }
 
+func (s *testMainSuite) TestSysSessionPoolGoroutineLeak(c *C) {
+	store := newStoreWithBootstrap(c, s.dbName+"goroutine_leak")
+	se, err := createSession(store)
+	c.Assert(err, IsNil)
+
+	// Test an issue that sysSessionPool doesn't call session's Close, cause
+	// asyncGetTSWorker goroutine leak.
+	before := runtime.NumGoroutine()
+	count := 200
+	var wg sync.WaitGroup
+	wg.Add(count)
+	for i := 0; i < count; i++ {
+		go func(se *session) {
+			_, _, err := se.ExecRestrictedSQL(se, "select * from mysql.user limit 1")
+			c.Assert(err, IsNil)
+			wg.Done()
+		}(se)
+	}
+	wg.Wait()
+	se.sysSessionPool().Close()
+	after := runtime.NumGoroutine()
+	c.Assert(after-before, Less, 3)
```

Comment:
It should be Equal, but I'm afraid this test disturbed by other factors that would make CI unstable, so I relax the strict check.